PoE HAT from Linux I2C

6by9 opened the pull request:
As discussed with pelwell by email, switching the PoE HAT fan control (initially) to Linux I2C so that it can co-exist with libcamera and KMS driving the 7" DSI panel.

I need to do the same to the PoE+ HAT power supply driver (shouldn't be a big job now), but am creating the PR for comments on any glaring structural issues.

6by9 committed:
mfd: simple-mfd-i2c: Add configuration for RPi POE HAT

The Raspbery Pi PoE+ HAT exposes a fan controller and power
supply status reporting via a single I2C address.

Create an MFD device that allows loading of the relevant
sub-drivers, with a shared I2C regmap.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
pwm: raspberrypi-poe: Add option of being created by MFD or FW

The firmware can only use I2C0 if the kernel isn't, therefore
with libcamera and DRM using it the PoE HAT fan control needs
to move to the kernel.

Add the option for the driver to be created by the PoE HAT core
MFD driver, and use the I2C regmap that provides to control fan
functions.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dtoverlays: Add option for PoE HAT to use Linux I2C instead of FW.

Adds parameter "i2c" to use the PoE HAT I2C MFD driver to instantiate
the PoE HAT drivers, instead of being off the firmware.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
defconfigs: Add MFD_RASPBERRYPI_POE_HAT to Pi defconfigs.

Required for PoE HAT when driven from Linux over I2C.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on arch/arm/boot/dts/overlays/rpi-poe-overlay.dts:
```diff
@@ -89,6 +79,52 @@
+	fragment@3 {
+		target = <&firmware>;
+		__overlay__ {
+			poe_pwm: pwm {
```
Including `fw` somewhere in the label name (as used to be the case) would be clearer.

6by9 replied:
Reverted the name. I'd thought I could change the thing pointed to via __overlay__ / __dormant__ so was looking for one generic name, but that doesn't work.

pelwell commented on arch/arm/boot/dts/overlays/rpi-poe-overlay.dts:
```diff
@@ -89,6 +79,52 @@
+				status = "okay";
+
+				poe_mfd_pwm: poe_pwm@0xf0 {
+					compatible = "raspberrypi,firmware-poe-pwm";
```
This compatible string is no longer correct in the mfd case. I'm not saying it has to change, but using two different compatible strings might help (if only for error checking).

6by9 replied:
OK, will add a second compatible, even though it'll do exactly the same thing.

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -2858,6 +2858,8 @@ Params: poe_fan_temp0           Temperature (in millicelcius) at which the fan
+                                the firmware.
 
 
 Name:   rpi-poe-plus
```
`rpi-poe-plus` inherits the `i2c` parameter, so also needs a README addition.

6by9 replied:
I was looking at PoE+ with the power driver as well, but yes.

pelwell commented on drivers/pwm/pwm-raspberrypi-poe.c:
```diff
@@ -52,34 +57,41 @@ static int raspberrypi_pwm_set_property(struct rpi_firmware *firmware,
+			ret = -EIO;
+	} else {
+		ret = regmap_write(pwm->regmap, pwm->offset + reg, val);
+		if (ret)
```
This is superfluous.

6by9 replied:
I'd already noted it in copying the same basis to the power driver. Done.

pelwell replied:
You missed this one.

pelwell commented on arch/arm/boot/dts/overlays/rpi-poe-overlay.dts:
```diff
@@ -89,6 +79,52 @@
+			#size-cells = <0>;
+			status = "okay";
+
+			poe@0x51 {
```
The preferred name is `poe@51`.

6by9 replied:
Not sure what I was copying that used the 0x. Ho hum.

pelwell commented on arch/arm/boot/dts/overlays/rpi-poe-overlay.dts:
```diff
@@ -89,6 +79,52 @@
+				reg = <0x51>;
+				status = "okay";
+
+				poe_mfd_pwm: poe_pwm@0xf0 {
```
The preferred name is `pwm@f0`.

pelwell reviewed (commented):
That's surprisingly compact. Minor comments only.

6by9 committed:
power: rpi-poe: Drop CURRENT_AVG as it is not hardware averaged

As documented the _AVG parameters are meant to be hardware
averaged, but the implementation for the PoE+ HAT was done in
software in the firmware.

Drop the property.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
power: rpi-poe: Add option of being created by MFD or FW

The firmware can only use I2C0 if the kernel isn't, therefore
with libcamera and DRM using it the PoE HAT fan control needs
to move to the kernel.

Add the option for the driver to be created by the PoE HAT core
MFD driver, and use the I2C regmap that provides to control fan
functions.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
OK, I think this is good for a proper review.
The fw change to add the i2c param automatically is up as a merge request too.

6by9 commented:
Updated to get rid of that duplicated ```return ret;``` (I can't find the comment for some reason).

pelwell commented on arch/arm/boot/dts/overlays/rpi-poe-plus-overlay.dts:
```diff
@@ -16,6 +16,21 @@
+		__dormant__ {
+			rpi-poe-power-supply@f2 {
+				compatible = "raspberrypi,rpi-poe-power-supply";
+				reg = <0xf0>;
```
Sorry - I just noticed this should be 0xf2.

6by9 replied:
Agreed, but then needs a slight rejig as RPI_POE_ADC_REG is currently 2, so that needs to be reset to 0, and FW needs a base offset to be added.

6by9 replied:
Done.
